src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -238,7 +238,10 @@ internal async Task<IServiceChannelDispatcher> GetAuthChannelDispatcher(IChannel
         /// <returns></returns>
         internal Task<IServiceChannelDispatcher> GetInnerServiceChannelDispatcher(IChannel outerChannel)
         {
-            return InnerServiceDispatcher.CreateServiceChannelDispatcherAsync(outerChannel);
+            lock (ThisLock)
+            {
```

Comment:
That method is async so the lock will only be held until the first await. You need to change ThisLock from object to AsyncLock and replace all usages of lock(ThisLock) to the using style of lock used with AsyncLock. Then you need to await the call on InnerDispatcher.CreateServiceChannelDispatcherAsync so that the lock is held for the duration of the method.

Reply:
let's  discuss